packages/block-editor/src/components/inner-blocks/use-nested-settings-update.js

```diff
@@ -77,10 +77,7 @@ export default function useNestedSettingsUpdate(
 	// otherwise if the arrays change length but the first elements are equal the comparison,
 	// does not works as expected.
 	const _allowedBlocks = useShallowMemo( allowedBlocks );
-
-	const _prioritizedInserterBlocks = useMemo(
-		() => prioritizedInserterBlocks,
-		// eslint-disable-next-line react-hooks/exhaustive-deps
+	const _prioritizedInserterBlocks = useShallowMemo(
```

Comment:
That's a good point 👍 

I've seen you working on it [in another repository](https://github.com/Automattic/wp-calypso/blob/5989a7552a5ab41e9e737233b031d0d7ab117bb9/client/lib/use-memo-compare/index.ts#L4-L17).

Feel free to open a PR, I think it can be improved separately.

Reply:
@jsnajdr the hook you're proposing is unsafe because you're reading/writing refs in render, which is forbidden by the React Gods (can mess things up in concurrent rendering trees). When doing render-time computations, you simply can't use refs.

However, many use cases for render-time refs can be replaced by derived state, which is a technique consisting of calling a setState at render time, which causes the render function to instantly re-run synchronously, skipping all effects and other work of the kind. This is fine, as long as the state is guaranteed to settle, otherwise it can result in an endless loop. Can't find it right now, but the technique is officially listed somewhere in the official, updated docs.

Running a component's render function is typically pretty much free, so it'd be rare if this technique is a problem for performance. Re-rendering is a tool, not a danger to avoid.

It'd be another story imo if you found a specific perf problem somewhere that is caused by excessive re-rendering, that can happen but anything earlier than that feels like premature optimization (or even worsening, as sometimes re-rendering is the better option).

That's my 2cents anyway :)

Reply:
Some context for the original `useShallowMemo` - https://github.com/WordPress/gutenberg/pull/53943#discussion_r1315133134.

Reply:
I made this about 3 years ago: https://codesandbox.io/p/sandbox/previous-vs-previous-value-t7t3y?file=%2Fsrc%2FApp.tsx

Reply:
Btw, instead of forking it could be composed like this:

```ts
const [current, setCurrent] = useState(0);
const previousValue = usePreviousValue(value);
if (compareFn(value, previousValue)) {
  setCurrent(n => n + 1);
}
function useMemo(computation, [current]);
```

Reply:
Btw, in @jsnajdr's example there's no memoization going on, the computation is executed every time. I might be missing something though.

Reply:
> you're reading/writing refs in render, which is forbidden by the React Gods (can mess things up in concurrent rendering trees)

Yes, forbidden by React Gods, we don't have any stronger argument than that. We don't know what exactly can mess up in concurrent rendering when reading/writing a ref this way.

When I look at how the `useMemo` hook is implemented, it's also reading/writing a ref-like value during render. It invokes the memo callback during render, and stores the value and the deps in exactly the same field (`hook.memoizedState`) where `useRef` stores its ref. So, if `useMemo` is concurrent-safe, then our ref-writing `useMemoCompare` hook is also safe.

> there's no memoization going on, the computation is executed every time.

That's because the hook doesn't have any dependencies, only the `compare` function. The goal of the hook is to not save computations, but to return an identical (`===`) object if possible.

Reply:
> I've seen you working on it in another repository.

Oh, that explains why I couldn't find it in the Gutenberg repo 🙂

Reply:
I feel like this convo is moving out of the original purpose of the PR. @DaniGuardiola @jsnajdr perhaps it makes sense to discuss this separately, in a PR that attempts to improve `useShallowMemo`? In the meantime, I think we can land this, as it's unrelated to the internal implementation of `useShallowMemo`. WDYT?

Reply:
The `useNestedSettingsUpdate` hook is used every time a container block has inner blocks, and that's a lot of usages. And while it's true that renders are cheap (not entirely free) it doesn't mean that all bets are off and we shouldn't care about rerenders at all. We've seen in #65564 that frequent rerenders can confuse the browser and can lead to broken UI.

This PR adds one more extra rerender to every `InnerBlocks` instance, it can take up to three renders before it settles down. And it's easily fixable.

Let's just update `useShallowMemo` to this implementation, it's been already tried elsewhere and it's very close to what `useMemo` is actually doing internally:
```js
function useShallowMemo( value ) {
  const valueRef = useRef( value );
  if ( ! isShallowEqual( valueRef.current, value ) ) {
    valueRef.current = value;
  }
  return valueRef.current;
}
```
The usual precautions about refs don't apply here. The function is idempotent, repeatedly updating with the same value has no effect. And the value returned by the hook is always the same. Not in the `===` sense, but in the semantic sense, regarding the information it contains.

Reply:
> Yes, forbidden by React Gods, we don't have any stronger argument than that. We don't know what exactly can mess up in concurrent rendering when reading/writing a ref this way.

BTW, React Compiler complains about that, too, so it will become a bigger issue if we want to adopt React 19.

Reply:
Can the React compiler error be fixed with this?
```js
function createMemo() {
  let value;
  return ( newValue ) => {
    if ( value === undefined || ! isShallowEqual( value, newValue ) {
      value = newValue;
    }
    return value;
  };
}

function useShallowMemo( value ) {
  const [ memo ] = useState( createMemo );
  return memo( value );
}
```
It doesn't use refs, it instead stores the value in an external custom object.

Reply:
Yes, React Compiler will no longer complain if we use this, assuming we fix the missed closing parenthesis from your example:

```diff
4c4
<     if ( value === undefined || ! isShallowEqual( value, newValue ) {
---
>     if ( value === undefined || ! isShallowEqual( value, newValue ) ) {
```

Reply:
Perfect, I propose to use it then 🙂 We have a per-instance `memo` function that remembers one last value.

Reply:
Sounds good - mind opening a separate PR for it? @DaniGuardiola could have some feedback 😅

Reply:
I think the discussion about `useShallowMemo` and the "right" implementation, whatever it turns out to be, naturally belongs to this PR. Otherwise, why add another usage of `useShallowMemo` if we're not sure about it?

Reply:
I'm good with the approach. @DaniGuardiola could you take a look at the approach?

Reply:
I disagree. Here we have a special case where we're implementing `useMemo` in user-space, with only slight modifications. Instead of doing a "is previous deps array shallowly equal to the next deps array" comparison, our comparison is a bit different, that's all. There is no scenario where tearing can happen.

If derived state was always the way to go, we wouldn't need `useMemo`.

Reply:
I think the Onus is on you to prove how this won't cause tearing. The rules are clear are the reasons behind them too. If you want to prove that this won't cause tearing, explain why.

I would also like to see the performance problem you're optimizing for by breaking known rules.

Finally, note that the React compiler warning is still there, in spirit, even if you managed to silence it by doing a clever scoped value/initial state/function trick that escapes what their linter can detect. Still reading/writing a ref on render, essentially. So I would like to know why exactly the warning can be ignored, because this is no different than an `// eslint-ignore` comment, except for the React compiler.

I'm personally not satisfied with "useMemo does it". The rules exist, and I don't have enough context on the internals of react to understand why things are implemented the way they are. They are internals (and I do not WANT_TO_BE_FIRED 😁).

Reply:
I will add that even if this happens to be safe today (which I don't think it is), there's nothing preventing the React team from releasing a version that finally breaks, because we're still breaking the public contract and rules.

It might even happen with the React compiler. Right now it's throwing a warning, but we're ignoring it which means it will actually compile anyway, and who knows if that will break the component due to reading/writing a ref on render.


I do not think any of this is worth it just to save the extra microseconds that a render (a.k.a. running a function, NOT effects, NOT reconciliation, NOT diffing, NOT applying changes to the DOM, just running a pretty lightweight function) takes.

Reply:
I think we're stuck in a disagreement here and I don't see any way out.

> If you want to prove that this won't cause tearing, explain why.

The proof is a straightforward syllogism:
1. `useMemo` doesn't cause tearing.
2. `useShallowMemo` is implemented exactly the same way as `useMemo`. Reading and writing the value that's stored in `hook.memoizedState`.
3. Therefore, `useShallowMemo` doesn't cause tearing.

> The rules exist, and I don't have enough context on the internals of react to understand why things are implemented the way they are.

The relevant internals are not hard to understand, just have a look at the `mountRef`, `updateRef`, `mountMemo` and `updateMemo` functions in `ReactFiberHooks.js` and not how they read and write the `hook.memoizedState` value.

Yes, there are rules, but here we have an exception from them.

Reply:
> I do not think any of this is worth it just to save the extra microseconds that a render takes.

It's not about one additional marginal render, but if we become completely careless about rerenders, they will multiply and will start causing problems. Every hook that "doesn't care" will add one rerender or more, and suddenly we have twenty.

In this case the extra render is not justified, it can be single-pass without any problems. We already have plenty of rerenders that need to be there because the component needs to update after its DOM element is created. Resize observer is an example. Also AriaKit is very liberal abound rerenders, e.g., a mere `<Button>` will render three times before it settles down, because it needs to check if the DOM element is a native `button`. Compounded together in a large component tree, this is not great.

Reply:
Well, do you have a benchmark, or maybe a user report that indicates a performance problem? So far we agree that there's an extra render, but I do not agree at all that it is a problem. I think we could do 100 more and it would still not be a problem. But I'm open to being proven wrong.

Anyway, it does seem that we're not getting any closer to agreement, so maybe we can rope other folks in to see what they think?

Reply:
I've given it a bit more thought, and while I see you both have points, I prefer going with the more pragmatic and less ideological way, which is to go with @jsnajdr's solution. Can we agree to disagree @DaniGuardiola?

Reply:
I don't agree that breaking the rules to prevent renders without a clear performance issue is the pragmatic and less ideological way, I'd argue it's the opposite. That said, we can agree to disagree on that too :P

Just hope I don't have a "told you so" moment a few months down the line. No hard feelings either way!

Reply:
_In the general case_, I do agree that we shouldn't be disobeying the React Gods without a substantial upside, and without a mechanism to catch problems when things start to break. Especially because this seems to be the type of thing where the kinds of bugs that could possibly occur in the future are very subtle.

I'm not privy enough of React internals to assess how much of a special exception this is to the general case. In the general case, just because an official hook adopts an implementation internally right now doesn't necessarily make it safe to imitate in userland. Could they rewrite the internal `useMemo` implementation without us noticing? Could they add a special safeguard for `useMemo` that doesn't apply to userland imitations? If there really is no magic and we can confidently answer No to those questions, then sure, maybe this is a special exception. (Unfortunately I cannot confidently answer those questions 😅)

Reply:
I think it's helpful to try to really understand why exactly reading/writing refs during render can be dangerous. The metaphors about rules written by gods are counterproductive, and also not completely innocent, because they suggest that we're dealing with some revealed truths that are inaccessible to reason. That's not true.

One example of a bad component is this:
```js
function Foo( { value } ) {
  const valueRef = useRef();
  valueRef.current = value;
  return <button onClick={ () => { console.log( valueRef.current ); } } />
}
```
Here, if you render `<Foo value="hello"/>`, clicking on the button will print `hello`. That's good. Now React can start rerendering the tree that contains `<Foo />`, with another value of the prop, but that render is aborted. It suspends or is interrupted, and the results are thrown away. DOM is not changed and effects are not run. But the `valueRef.current = value` modification was done and it stays. It's an assignment very similar to calling `this.value = ...` on a component instance, there is nothing sophisticated about it.

This means that an aborted render of `<Foo value="world" />` will actually modify `valueRef` and clicking on the button will log `world`. The modification that was not supposed to happen has a material effect on the rendered UI. We can construct many other examples like this where aborted, tentative renders can leak into the active UI. That's why the ref access should be moved into effects, because effects run only when the result of the render was actually commited. Or we should use state, because the state of a committed component is isolated from the state of the tentative render, and the updates are not done directly, but go though a managed queue. Refs are not isolated or managed like this, they are "naked" values.

But not all ref reads/writes are dangerous like this. The [React docs contain this example](https://react.dev/reference/react/useRef#avoiding-recreating-the-ref-contents) that gives us a first piece of evidence that the rules are not absolute.
```js
function Video() {
  const playerRef = useRef(null);
  if (playerRef.current === null) {
    playerRef.current = new VideoPlayer();
  }
  // ...
```
Here we read and write a ref during render, but the docs say this is fine!

> Normally, writing or reading `ref.current` during render is not allowed. However, it’s fine in this case because the result is always the same, and the condition only executes during initialization so it’s fully predictable.

Indeed, if a tentative render that's going to be aborted initializes the `VideoPlayer`, we don't mind, because we would create it the same way anyway. And most of the time the tentative render will find out the `VideoPlayer` already exists and will do nothing. Nothing bad can happen.

Now let's check if `useMemo` could also possibly be safe like this. Let's implement `useMemo` in user space:
```js
function useMemo( create, deps ) {
  const prevState = useRef();
  if ( prevState.current === undefined || ! areDepsEqual( prevState.current[ 1 ], deps ) ) {
    prevState.current = [ create(), deps ];
  }
  return prevState.current[ 0 ];
}
```
Two things to note here:
1. The hook's `create` parameter is always capable to produce a valid return value. And it's constructed from "pure" React values: the props, the state, the context. Also the `deps` are derived from these "pure" values.
2. The hook is very skeptical about the value in the `prevState` ref. Before using it, it carefully verifies that it's valid, i.e., whether it was constructed with the same `deps`. If some bad value leaked into the `prevState` ref from an aborted render, this will be detected, the value will be ignored, and will be recalculated from the "pure" `create` function.

That means that the `useMemo` hook is resilient against bad values that leak into it from aborted renders. The only thing that the aborted renders can ever achieve is that they invalidate your memoization and force you to recreate it, but the hook will never return a bad value.

That's why the native `useMemo` hook is safe, and also why a user-space `useMemo` or `useShallowMemo` implementation is also safe if done well.

Reply:
@jsnajdr Thanks for the explanation!

Reply:
Thanks y'all for the feedback!

I agree with @jsnajdr and will move to ship this, and will keep an eye in the future.